h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+  */
+  updateDomainModel = function(domainModel, viewModel) {
+    var i;
+    var tagTexts = [];
+    for (i = 0; i < viewModel.tags.length; i++) {
+      tagTexts.concat(viewModel.tags[i].text);
+    }
+    angular.extend(domainModel, viewModel, {tag: tagTexts});
+  };
+
+  /**
+   * Create or update the existing draft for this annotation using
+   * the text and tags from the domain model in `draft`.
+   */
+  updateDraft = function(draft) {
+    // Drafts only preserve the text, tags and permissions of the annotation
+    // (i.e. only the bits that the user can edit), changes to other
+    // properties are not preserved.
+    var changes = {};
+    if (draft.text) {
+      changes.text = draft.text;
+    }
+    if (draft.tags) {
+      changes.tags = draft.tags;
+    }
+    if (draft.permissions) {
+      changes.permissions = draft.permissions;
+    }
+    drafts.update(model, changes);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#save
+    * @description Saves any edits and returns to the viewer.
+    */
+  vm.save = function() {
+    var newTags;
+    var onFulfilled;
+    var onRejected;
+    var updatedModel;
+
+    if (!model.user) {
+      return flash.info('Please sign in to save your annotations.');
+    }
+
+    if (!validate(vm.annotation)) {
+      return flash.info('Please add text or a tag before publishing.');
+    }
+
+    // Update stored tags with the new tags of this annotation.
+    newTags = vm.annotation.tags.filter(function(tag) {
+      var tags = model.tags || [];
+      return (tags.indexOf(tag.text) === -1);
+    });
+    tags.store(newTags);
+
+    switch (vm.action) {
+      case 'create':
+        updateDomainModel(model, vm.annotation);
+        onFulfilled = function() {
+          $rootScope.$emit('annotationCreated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return model.$create().then(onFulfilled, onRejected);
+
+      case 'edit':
+        updatedModel = angular.copy(model);
+        updateDomainModel(updatedModel, vm.annotation);
+        onFulfilled = function() {
+          angular.copy(updatedModel, model);
+          $rootScope.$emit('annotationUpdated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return updatedModel.$update({
+          id: updatedModel.id
+        }).then(onFulfilled, onRejected);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#reply
+    * @description
+    * Creates a new message in reply to this annotation.
+    */
+  vm.reply = function() {
+    var id = model.id;
+    var references = model.references || [];
+    var reply;
+    var uri = model.uri;
+
+    if (typeof references === 'string') {
```

Comment:
Agree - perhaps you could make a note to check though. We might be able to simplify this post-migration.